### PR TITLE
kubeadm: add support for parallel image pulls in v1beta4

### PR DESCRIPTION
**Pull request by neolit123**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

add support for parallel image pulls in v1beta4 via the NodeRegistration.ImagePullSerial.
default value is true - serial is default.
upgrade configuration must be updated seperately.

#### Which issue(s) this PR fixes:

xref https://github.com/kubernetes/kubeadm/issues/2780

#### Special notes for your reviewer:

example output

```
I0106 14:32:54.555064   16383 checks.go:832] using image pull policy: IfNotPresent
W0106 14:32:54.641088   16383 checks.go:847] detected that the sandbox image "registry.k8s.io/pause:3.6" of the container runtime is inconsistent with that used by kubeadm. It is recommended to use "registry.k8s.io/pause:3.9" as the CRI sandbox image.
I0106 14:32:54.641175   16383 runtime.go:155] pulling all images in parallel
I0106 14:32:54.691572   16383 runtime.go:166] image exists: registry.k8s.io/kube-scheduler:v1.29.0
I0106 14:32:58.204952   16383 runtime.go:175] done pulling: registry.k8s.io/pause:3.9
I0106 14:33:10.432991   16383 runtime.go:175] done pulling: registry.k8s.io/coredns/coredns:v1.11.1
I0106 14:33:31.178267   16383 runtime.go:175] done pulling: registry.k8s.io/kube-proxy:v1.29.0
I0106 14:33:33.189903   16383 runtime.go:175] done pulling: registry.k8s.io/etcd:3.5.11-0
I0106 14:33:35.300818   16383 runtime.go:175] done pulling: registry.k8s.io/kube-controller-manager:v1.29.0
I0106 14:33:35.430317   16383 runtime.go:175] done pulling: registry.k8s.io/kube-apiserver:v1.29.0

```

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by pacoxu** [comment]
Should we support experimental api for `kubeadm config images`?

```
[root@daocloud ~]# ./kubeadm config images list --config k.yaml
could not convert cfg to an internal cfg: experimental API spec: "kubeadm.k8s.io/v1beta4" is not allowed. You can use the --allow-experimental-api flag if the command supports it.
To see the stack trace of this error execute with --v=5 or higher
```

**Comment by neolit123** [comment]
/retest pull-kubernetes-linter-hints 
prow pod failed

**Comment by neolit123** [comment]
> Should we support experimental api for `kubeadm config images`?
> 
> ```
> [root@daocloud ~]# ./kubeadm config images list --config k.yaml
> could not convert cfg to an internal cfg: experimental API spec: "kubeadm.k8s.io/v1beta4" is not allowed. You can use the --allow-experimental-api flag if the command supports it.
> To see the stack trace of this error execute with --v=5 or higher
> ```

we could technically add the --allow-experimental-api flag to any command that already has --config, IIUC.
but i think we need to understand what we win from that.

kubeadm users are unlikely to test the experimental APIs and give us feedback. kinder does not support multiple APIs (currently only supports v1beta3). so it seems it would only benefit us, the developers for testing our changes. i can already patch kubeadm to test it locally.

what we don't want is users to start using an experimental API.

**Comment by pacoxu** [comment]
> what we don't want is users to start using an experimental API.

`--allow-experimental-api   Allow validation/migration to experimental, unreleased APIs.` 
Only benefit testing as we now only can migrate and validate the new APIs. 

If kinder doesn't support, when are we ready to upgrade to v1beta4 without enough testing with all CI jobs?

**Comment by neolit123** [comment]
> > what we don't want is users to start using an experimental API.
> 
> `--allow-experimental-api Allow validation/migration to experimental, unreleased APIs.` Only benefit testing as we now only can migrate and validate the new APIs.
> 
> If kinder doesn't support, when are we ready to upgrade to v1beta4 without enough testing with all CI jobs?

basically we can't move kinder to v1beta4 before v1beta4 is possible to be passed as --config everywhere.
we could add --allow-experimental-api to all commands and integrate kinder to support this flag.

it seems like the safer option, but it's also a lot of work.

**Comment by pacoxu** [comment]
Overall LGTM(a quick review)
- key change is https://github.com/kubernetes/kubernetes/pull/122616/commits/1c5773ca9100cc633fc0fd21bdca8d74aabb2f79 and https://github.com/kubernetes/kubernetes/pull/122616/files#diff-831431801d11a8ecc1116374c90a45c20e4ff67824003f430f588e07b994edb6R851-R856.

File: cmd/kubeadm/app/util/runtime/runtime.go [hunk 1]
```diff
@@ -139,6 +140,51 @@ func (runtime *CRIRuntime) PullImage(image string) error {
...
+				err = errors.WithMessagef(err, "failed to pull image %s", image)
+			} else {
+				klog.V(1).Infof("done pulling: %s", image)
+			}
+			errChan <- err
+		}()
+	}
+
+	for i := 0; i < len(images); i++ {
+		errs = append(errs, <-errChan)
```
Comment (SataQiu) [inline cmd/kubeadm/app/util/runtime/runtime.go:182]: Do we need to check if the err is nil?
```suggestion
                if err := <-errChan; err != nil {
                     errs = append(errs, err)
                }
```

Comment (neolit123) [reply cmd/kubeadm/app/util/runtime/runtime.go]: updated.

**Comment by neolit123** [comment]
rebased. looking for another review pass / LGTM.

**Pull request merged**